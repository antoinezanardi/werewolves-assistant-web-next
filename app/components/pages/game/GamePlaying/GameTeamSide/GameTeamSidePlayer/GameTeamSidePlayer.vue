<template>
  <GlowCapture>
    <GlowElement>
      <div
        id="game-team-side-player"
        class="border-2 border-gray-700 p-2 rounded-md w-full"
        :class="playerSideGlowClass"
        :data-testid="`game-team-side-player-${player.name}`"
      >
        <div class="flex justify-center">
          <NuxtImg
            v-if="!player.isAlive"
            v-p-tooltip="$t('components.GameTeamSidePlayer.thisPlayerIsDead')"
            :alt="$t('components.GameTeamSidePlayer.thisPlayerIsDead')"
            class="me-2"
            height="20"
            src="/svg/game/player/dead.svg"
            width="20"
          />

          <div
            id="player-name"
            class="text-center truncate"
            :class="{ 'line-through decoration-1': !player.isAlive }"
          >
            {{ player.name }}
          </div>
        </div>

        <VuePrimeDivider class="!my-1"/>

        <small
          id="player-role-name"
          class="flex justify-center"
        >
          {{ $t(`shared.role.name.${player.role.current}`) }}
        </small>

        <div class="flex items-center mt-2">
          <RoleImage
            v-if="player.side.current === 'werewolves'"
            id="player-werewolf-role-image"
            :alt="roleImageAlt"
            class="!border-2"
            :class="roleImageClasses"
            definition="small"
            :role-name="player.role.current"
            sizes="50"
          />

          <div
            id="player-attributes"
            class="grid grid-cols-3 grow mx-2"
          >
            <GameTeamSidePlayerAttribute
              v-for="attribute in player.attributes"
              :key="attribute.name"
              :attribute="attribute"
              class="p-1"
              :player="player"
            />
          </div>

          <RoleImage
            v-if="player.side.current === 'villagers'"
            id="player-villager-role-image"
            :alt="roleImageAlt"
            class="!border-2"
            :class="roleImageClasses"
            definition="small"
            :role-name="player.role.current"
          />
        </div>
      </div>
    </GlowElement>
  </GlowCapture>
</template>

<script lang="ts" setup>
import type { GameTeamSidePlayerProps } from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/game-team-side-player.types";
import GameTeamSidePlayerAttribute from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/GameTeamSidePlayerAttribute/GameTeamSidePlayerAttribute.vue";
import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";

const props = defineProps<GameTeamSidePlayerProps>();

const { t } = useI18n();

const roleImageAlt = computed<string>(() => t("components.GameTeamSidePlayer.roleImageAlt", { playerName: props.player.name }));

const playerSideGlowClass = computed<string>(() => {
  const { player } = props;
  if (!player.isAlive) {
    return "";
  }
  return player.side.current === "werewolves" ? "glow:border-red-500" : "glow:border-green-500";
});

const roleImageClasses = computed<string>(() => {
  const { player } = props;
  let classes = playerSideGlowClass.value;
  if (!player.isAlive) {
    classes += " grayscale";
  }
  return classes;
});
</script>