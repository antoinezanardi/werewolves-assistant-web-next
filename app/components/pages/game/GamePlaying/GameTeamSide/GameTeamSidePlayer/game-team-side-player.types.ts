import type { Player } from "~/composables/api/game/types/players/player.class";

type GameTeamSidePlayerProps = {
  player: Player;
};

export type { GameTeamSidePlayerProps };