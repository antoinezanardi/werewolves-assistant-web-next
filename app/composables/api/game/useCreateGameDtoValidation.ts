import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };