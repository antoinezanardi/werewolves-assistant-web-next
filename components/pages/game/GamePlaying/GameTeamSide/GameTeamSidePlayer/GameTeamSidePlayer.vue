<template>
  <div id="game-team-side-player">
    <div
      id="player-name"
      class="text-center text-truncate"
    >
      {{ player.name }}
    </div>

    <hr class="my-1">

    <small
      id="player-role-name"
      class="flex justify-center"
    >
      {{ $t(`shared.role.name.${player.role.current}`) }}
    </small>

    <div class="flex mt-1">
      <RoleImage
        v-if="player.side.current === RoleSides.WEREWOLVES"
        id="player-werewolf-role-image"
        :alt="roleImageAlt"
        definition="small"
        height="50"
        :role-name="player.role.current"
        width="50"
      />

      <div
        id="player-attributes"
        class="flex grow"
      />

      <RoleImage
        v-if="player.side.current === RoleSides.VILLAGERS"
        id="player-villager-role-image"
        :alt="roleImageAlt"
        definition="small"
        height="50"
        :role-name="player.role.current"
        width="50"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import type { GameTeamSidePlayerProps } from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/game-team-side-player.types";
import RoleImage from "~/components/shared/role/RoleImage/RoleImage.vue";
import { RoleSides } from "~/composables/api/role/enums/role.enums";

const props = defineProps<GameTeamSidePlayerProps>();

const { t } = useI18n();

const roleImageAlt = computed<string>(() => t("components.GameTeamSidePlayer.roleImageAlt", { playerName: props.player.name }));
</script>

<style lang="scss" scoped>
#game-team-side-player {
  border: 1px solid #4d4d4d;
  border-radius: 10px;
  padding: 5px;
  width: 100%;
}
</style>