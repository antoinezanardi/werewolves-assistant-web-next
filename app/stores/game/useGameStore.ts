import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };