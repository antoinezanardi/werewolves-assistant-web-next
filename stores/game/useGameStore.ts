import type { AsyncDataRequestStatus } from "#app/composables/asyncData";
import { defineStore } from "pinia";

import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const { getGame: fetchGameFromApi } = useFetchGames();

  const game = ref<Game>(new Game());
  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }
  return {
    game,
    fetchingGameStatus,
    fetchAndSetGame,
  };
});

export { useGameStore };