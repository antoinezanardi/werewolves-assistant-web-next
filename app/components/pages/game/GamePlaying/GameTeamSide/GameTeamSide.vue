<template>
  <div
    id="game-playground-team-side"
    class="flex flex-col h-full"
  >
    <div class="flex items-center justify-evenly">
      <NuxtImg
        :alt="$t(`components.GameTeamSide.sideAltText.${props.side}`)"
        height="50"
        placeholder="/svg/misc/ripples.svg"
        placeholder-class="w-50 h-50"
        :src="sideSvgIconPath"
        width="50"
      />

      <div>
        <h2
          id="side-title"
          class="my-1 text-center"
        >
          {{ $t(`shared.role.side.${side}`) }}
        </h2>

        <h4
          id="alive-team-players"
          class="my-1 text-center"
        >
          {{ alivePlayersText }}
        </h4>
      </div>

      <div/>
    </div>

    <PrimeVueDivider/>

    <div
      id="team-players"
      class="flex flex-col grow overflow-y-auto px-1 scrollbar-thin"
    >
      <GameTeamSidePlayer
        v-for="player in aliveFirstTeamPlayers"
        :key="player._id"
        class="my-1"
        :player="player"
      />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { storeToRefs } from "pinia";

import type { GameTeamSideProps } from "~/components/pages/game/GamePlaying/GameTeamSide/game-team-side.types";
import GameTeamSidePlayer from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/GameTeamSidePlayer.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameStore } from "~/stores/game/useGameStore";

const props = defineProps<GameTeamSideProps>();

const { t } = useI18n();

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const teamPlayers = computed<Player[]>(() => game.value.players.filter(player => player.side.current === props.side));

const aliveFirstTeamPlayers = computed<Player[]>(() => teamPlayers.value.toSorted((playerA, playerB) => {
  if (playerA.isAlive && !playerB.isAlive) {
    return -1;
  }
  return 1;
}));

const alivePlayersText = computed<string>(() => {
  const alivePlayers = teamPlayers.value.filter(player => player.isAlive);

  return t(`components.GameTeamSide.aliveTeamPlayers`, {
    alivePlayers: alivePlayers.length,
    totalPlayers: teamPlayers.value.length,
  });
});

const sideSvgIconPath = computed<string>(() => {
  if (props.side === "villagers") {
    return "/svg/role/villager.svg";
  }
  return "/svg/role/werewolf.svg";
});
</script>