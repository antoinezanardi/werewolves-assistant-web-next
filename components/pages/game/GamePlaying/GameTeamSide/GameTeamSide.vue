<template>
  <div id="game-playground-team-side">
    <GameTeamSidePlayer
      v-for="player in teamPlayers"
      :key="player._id"
      :player="player"
    />
  </div>
</template>

<script lang="ts" setup>
import { storeToRefs } from "pinia";

import type { GameTeamSideProps } from "~/components/pages/game/GamePlaying/GameTeamSide/game-team-side.types";
import GameTeamSidePlayer from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/GameTeamSidePlayer.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { useGameStore } from "~/stores/game/useGameStore";

const props = defineProps<GameTeamSideProps>();

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const teamPlayers = computed<Player[]>(() => game.value.players.filter(player => player.side.current === props.side));
</script>