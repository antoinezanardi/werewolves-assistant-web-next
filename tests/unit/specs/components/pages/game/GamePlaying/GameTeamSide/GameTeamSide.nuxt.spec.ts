import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";

import type { NuxtImg } from "#components";
import type { GameTeamSideProps } from "~/components/pages/game/GamePlaying/GameTeamSide/game-team-side.types";
import GameTeamSide from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSide.vue";
import GameTeamSidePlayer from "~/components/pages/game/GamePlaying/GameTeamSide/GameTeamSidePlayer/GameTeamSidePlayer.vue";
import type { Player } from "~/composables/api/game/types/players/player.class";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameStore } from "~/stores/game/useGameStore";
import { createFakeGame } from "~/tests/unit/utils/factories/composables/api/game/game.factory";
import { createFakeAccursedWolfFatherAlivePlayer, createFakeBigBadWolfAlivePlayer, createFakeHunterAlivePlayer, createFakeSeerAlivePlayer, createFakeVillagerAlivePlayer, createFakeWerewolfAlivePlayer, createFakeWhiteWerewolfAlivePlayer } from "~/tests/unit/utils/factories/composables/api/game/player/player-with-role.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Game Team Side Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameTeamSide>>;
  const defaultProps: GameTeamSideProps = { side: "villagers" };
  const testingPinia = { initialState: { [StoreIds.GAME]: { game: createFakeGame() } } };

  async function mountGameTeamSideComponent(options: ComponentMountingOptions<typeof GameTeamSide> = {}): Promise<ReturnType<typeof mount<typeof GameTeamSide>>> {
    return mountSuspendedComponent(GameTeamSide, {
      props: options.props ?? defaultProps,
      global: { plugins: [createTestingPinia(testingPinia)] },
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameTeamSideComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Team Side Image", () => {
    it("should render the werewolves image when the side is werewolves.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "werewolves" } });
      const image = wrapper.findComponent<typeof NuxtImg>("[alt='Werewolves side']");

      expect(image.attributes("src")).toBe("/svg/role/werewolf.svg");
    });

    it("should render the villagers image when the side is villagers.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "villagers" } });
      const image = wrapper.findComponent<typeof NuxtImg>("[alt='Villagers side']");

      expect(image.attributes("src")).toBe("/svg/role/villager.svg");
    });
  });

  describe("Title", () => {
    it("should render the werewolves title when the side is werewolves.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "werewolves" } });
      const title = wrapper.find<HTMLHeadingElement>("#side-title");

      expect(title.text()).toBe("Werewolves");
    });

    it("should render the villagers title when the side is villagers.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "villagers" } });
      const title = wrapper.find<HTMLHeadingElement>("#side-title");

      expect(title.text()).toBe("Villagers");
    });
  });

  describe("Alive text", () => {
    it("should render the number of alive werewolves when the side is werewolves.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "werewolves" } });
      const gameStore = useGameStore();
      gameStore.game = createFakeGame({
        players: [
          createFakeWerewolfAlivePlayer(),
          createFakeBigBadWolfAlivePlayer({ isAlive: false }),
          createFakeVillagerAlivePlayer(),
          createFakeAccursedWolfFatherAlivePlayer(),
          createFakeWhiteWerewolfAlivePlayer(),
        ],
      });
      await nextTick();
      const aliveText = wrapper.find<HTMLSpanElement>("#alive-team-players");

      expect(aliveText.text()).toBe("components.GameTeamSide.aliveTeamPlayers, {\"alivePlayers\":3,\"totalPlayers\":4}");
    });

    it("should render the number of alive villagers when the side is villagers.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "villagers" } });
      const gameStore = useGameStore();
      gameStore.game = createFakeGame({
        players: [
          createFakeWerewolfAlivePlayer(),
          createFakeSeerAlivePlayer(),
          createFakeVillagerAlivePlayer(),
          createFakeAccursedWolfFatherAlivePlayer({ isAlive: false }),
          createFakeHunterAlivePlayer({ isAlive: false }),
        ],
      });
      await nextTick();
      const aliveText = wrapper.find<HTMLSpanElement>("#alive-team-players");

      expect(aliveText.text()).toBe("components.GameTeamSide.aliveTeamPlayers, {\"alivePlayers\":2,\"totalPlayers\":3}");
    });
  });

  describe("Game Team Side Players", () => {
    it("should render the werewolves players with alive players first when the side is werewolves.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "werewolves" } });
      const gameStore = useGameStore();
      gameStore.game = createFakeGame({
        players: [
          createFakeWerewolfAlivePlayer(),
          createFakeBigBadWolfAlivePlayer(),
          createFakeVillagerAlivePlayer(),
          createFakeAccursedWolfFatherAlivePlayer({ isAlive: false }),
          createFakeWhiteWerewolfAlivePlayer(),
        ],
      });
      await nextTick();
      const players = wrapper.findAllComponents<typeof GameTeamSidePlayer>(GameTeamSidePlayer);

      expect(players).toHaveLength(4);
      expect(players[0].props("player")).toStrictEqual<Player>(gameStore.game.players[0]);
      expect(players[1].props("player")).toStrictEqual<Player>(gameStore.game.players[1]);
      expect(players[2].props("player")).toStrictEqual<Player>(gameStore.game.players[4]);
      expect(players[3].props("player")).toStrictEqual<Player>(gameStore.game.players[3]);
    });

    it("should render the villagers players with alive players first when the side is villagers.", async() => {
      wrapper = await mountGameTeamSideComponent({ props: { side: "villagers" } });
      const gameStore = useGameStore();
      gameStore.game = createFakeGame({
        players: [
          createFakeWerewolfAlivePlayer(),
          createFakeSeerAlivePlayer({ isAlive: false }),
          createFakeVillagerAlivePlayer(),
          createFakeAccursedWolfFatherAlivePlayer(),
          createFakeHunterAlivePlayer(),
        ],
      });
      await nextTick();
      const players = wrapper.findAllComponents<typeof GameTeamSidePlayer>(GameTeamSidePlayer);

      expect(players).toHaveLength(3);
      expect(players[0].props("player")).toStrictEqual<Player>(gameStore.game.players[2]);
      expect(players[1].props("player")).toStrictEqual<Player>(gameStore.game.players[4]);
      expect(players[2].props("player")).toStrictEqual<Player>(gameStore.game.players[1]);
    });
  });
});