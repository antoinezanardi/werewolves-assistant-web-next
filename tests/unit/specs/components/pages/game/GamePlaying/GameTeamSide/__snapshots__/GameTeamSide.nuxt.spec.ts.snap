// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Team Side Component > should match snapshot when rendered. 1`] = `
"<div id="game-playground-team-side" class="flex flex-col h-full">
  <div class="flex items-center justify-evenly">
    <nuxt-img-stub alt="Villagers side" height="50" src="/svg/role/villager.svg" width="50" preload="false" crossorigin="false" placeholder="/svg/misc/infinite-spinner.svg"></nuxt-img-stub>
    <div>
      <h2 id="side-title" class="my-1 text-center">Villagers</h2>
      <h4 id="alive-team-players" class="my-1 text-center">components.GameTeamSide.aliveTeamPlayers, {"alivePlayers":0,"totalPlayers":0}</h4>
    </div>
    <div></div>
  </div>
  <divider-stub></divider-stub>
  <div id="team-players" class="flex flex-col grow overflow-y-auto"></div>
</div>"
`;
